### PR TITLE
Retention policies

### PR DESCRIPTION
**Pull request by carycheng**

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @carycheng, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

File: boxsdk/client/client.py [hunk 1]
```diff
@@ -13,8 +13,11 @@
 from ..object.file import File
 from ..object.group import Group
 from ..object.group_membership import GroupMembership
+from ..object.retention_policy import RetentionPolicy
+from ..object.retention_policy_assignment import RetentionPolicyAssignment
 from ..util.shared_link import get_shared_link_header
 from ..util.translator import Translator
+from ..pagination.marker_based_object_collection import MarkerBasedObjectCollection
```
Comment (anthonywee) [inline boxsdk/client/client.py]: minor: alphabetical imports

File: boxsdk/client/client.py [hunk 2]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
         """
         # pylint:disable=no-self-use
         return self._session.get_url(endpoint, *args)
+
+
```
Comment (anthonywee) [inline boxsdk/client/client.py]: white space

File: boxsdk/client/client.py [hunk 3]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+
+    def retention_policy(self, retention_id):
+        """
+        Initialize a :class: `RetentionPolicy` object, whose box id is retention_id.
```
Comment (anthonywee) [inline boxsdk/client/client.py]: extra space between ":class:" and `RetentionPolicy`

File: boxsdk/client/client.py [hunk 4]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+        :rtype:
+            :class:`RetentionPolicy`
+        """
+        return RetentionPolicy(session=self._session, object_id=retention_id)
+
+
+
+    def retention_policy_assignment(self, assignment_id):
+        """
+        Initialize a :class: `RetentionPolicyAssignment` object, whose box id is assignment_id.
```
Comment (anthonywee) [inline boxsdk/client/client.py]: extra space after :class:

File: boxsdk/client/client.py [hunk 5]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+
+        :param assignment_id:
+            The box ID of the :class:`RetentionPolicyAssignment` object.
+        :type assignment_id:
+            `unicode`
+        :return:
+            A :class: `RetentionPolicyAssignment` object with the given entry ID.
```
Comment (anthonywee) [inline boxsdk/client/client.py]: extra space

File: boxsdk/client/client.py [hunk 6]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+        :type retention_id:
+            `unicode`
+        :return:
+            A :class: `RetentionPolicy` object with the given entry ID.
+        :rtype:
+            :class:`RetentionPolicy`
+        """
+        return RetentionPolicy(session=self._session, object_id=retention_id)
+
+
```
Comment (anthonywee) [inline boxsdk/client/client.py]: white space

File: boxsdk/client/client.py [hunk 7]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+    def retention_policy(self, retention_id):
+        """
+        Initialize a :class: `RetentionPolicy` object, whose box id is retention_id.
+
+        :param retention_id:
+            The box ID of the :class:`RetentionPolicy` object.
+        :type retention_id:
+            `unicode`
+        :return:
+            A :class: `RetentionPolicy` object with the given entry ID.
```
Comment (anthonywee) [inline boxsdk/client/client.py]: extra space

File: boxsdk/client/client.py (hunk 5 above)
Comment (anthonywee) [reply boxsdk/client/client.py]: "given assignment id"?

File: boxsdk/client/client.py [hunk 8]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+        """
+        return RetentionPolicyAssignment(session=self._session, object_id=assignment_id)
+
+
+    def create_retention_policy(
+            self,
+            policy_name,
+            policy_type,
+            disposition_action,
+            can_owner_extend_retention=False,
```
Comment (anthonywee) [inline boxsdk/client/client.py]: tabbing

File: boxsdk/client/client.py [hunk 9]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+            are_owners_notified=False,
+            retention_length=None,
+            custom_notification_recipients=None
```
Comment (anthonywee) [inline boxsdk/client/client.py]: trailing comma

File: boxsdk/client/client.py [hunk 10]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+        :return:
+            The newly created Retention Policy
+        :rtype:
+            :class:`RetentionPolicy`
+        """
+        url = '{0}/retention_policies'.format(API.BASE_API_URL)
+        retention_attributes = {
+            'policy_name': policy_name,
+            'policy_type': policy_type,
+            'disposition_action': disposition_action
```
Comment (anthonywee) [inline boxsdk/client/client.py]: trailing comma

File: boxsdk/client/client.py [hunk 11]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+            retention_attributes['are_owners_notified'] = are_owners_notified
+        if retention_length is not None:
+            retention_attributes['retention_length'] = retention_length
+        if custom_notification_recipients is not None:
+            retention_attributes['custom_notification_recipients'] = custom_notification_recipients
+        box_response = self._session.post(url, data=json.dumps(retention_attributes))
+        response = box_response.json()
+        return RetentionPolicy(self._session, response['id'], response)
+
+
```
Comment (anthonywee) [inline boxsdk/client/client.py]: white space

File: boxsdk/client/client.py [hunk 12]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+    def retention_policies(
+            self,
```
Comment (anthonywee) [inline boxsdk/client/client.py]: tabbing

File: boxsdk/client/client.py [hunk 13]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+            `unicode` or None
+        :param policy_type:
+            Set to either `finite` or `indefinite`
+        :type policy_type:
+            `unicode` or None
+        :param created_by_user_id:
+            A user id to filter the retention policies.
+        :type created_by_user_id:
+            `unicode` or None
+	    :param limit:
```
Comment (anthonywee) [inline boxsdk/client/client.py]: tabbing

File: boxsdk/client/client.py [hunk 14]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+            self,
+            policy_name=None,
+            policy_type=None,
+            created_by_user_id=None,
+            limit=None,
+            marker=None,
+            fields=None,
+    ):
+        """
+        Get the entries in the retention policy using limit-offset paging.
```
Comment (anthonywee) [inline boxsdk/client/client.py]: marker based paging?

File: boxsdk/client/client.py [hunk 15]
```diff
@@ -391,3 +394,159 @@ def get_url(self, endpoint, *args):
...
+            'created_by_user_id': created_by_user_id
+        }
+        return MarkerBasedObjectCollection(
+            session=self._session,
+            url=self._session.get_url('retention_policies'),
+            additional_params=additional_params,
+            limit=limit,
+            marker=marker,
+            fields=fields,
+            return_full_pages=False
```
Comment (anthonywee) [inline boxsdk/client/client.py]: trailing comma

File: boxsdk/object/base_api_json_object.py [hunk 16]
```diff
@@ -134,3 +134,12 @@ def response_object(self):
...
         return copy.deepcopy(self._response_object)
+
+    @property
+    def object_type(self):
+        """Return the Box type for the object.
+ 
+        :rtype:
+            `unicode`
+        """
+        return self._item_type
```
Comment (anthonywee) [inline boxsdk/object/base_api_json_object.py]: newline

File: boxsdk/object/retention_policy.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
+# coding: utf-8
+
+from __future__ import unicode_literals
```
Comment (anthonywee) [inline boxsdk/object/retention_policy.py]: Should also import `absolute_import`

File: boxsdk/object/retention_policy.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+from ..config import API
+from ..pagination.marker_based_object_collection import MarkerBasedObjectCollection
+
+
+class RetentionPolicy(BaseObject):
+    """Represents a Box retention policy."""
+
+    _item_type = 'retention_policy'
+
+    def get_url(self, *args): return self._session.get_url('retention_policies', self._object_id, *args)
```
Comment (anthonywee) [inline boxsdk/object/retention_policy.py]: doc string

File: boxsdk/object/retention_policy.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+            List of fields to request.
+        :type fields:
+            `Iterable` of `unicode`
+        """
+        url = self._session.get_url('retention_policy_assignments')
+        body = {
+            'policy_id': self.object_id,
+            'assign_to': {
+                'type': item.object_type,
+                'id': item.object_id
```
Comment (anthonywee) [inline boxsdk/object/retention_policy.py]: trailing comma

File: boxsdk/object/retention_policy.py [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+            }
```
Comment (anthonywee) [inline boxsdk/object/retention_policy.py]: trailing comma

File: boxsdk/object/retention_policy.py [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        response = self._session.post(url, data=json.dumps(body)).json()
+        return Translator().translate(response['type'])(
+            self._session,
+            response['id'],
+            response,
+        )
+
+
```
Comment (anthonywee) [inline boxsdk/object/retention_policy.py]: white space

File: boxsdk/object/retention_policy.py [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+        :type marker:
+            `unicode`
+        :param fields:
+            List of fields to request.
+        :type fields:
+            `Iterable` of `unicode`
+        :returns:
+            A list of assignments in the retention policy.
+        :rtype:
+            `list` of :class:`RetentionPolicyAssignment`
```
Comment (anthonywee) [inline boxsdk/object/retention_policy.py]: Fix rtype

File: boxsdk/object/retention_policy.py [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+            'type': assignment_type,
+        }
+        return MarkerBasedObjectCollection(
+            session=self._session,
+            url='{0}/retention_policies/{1}/assignments'.format(API.BASE_API_URL, self.object_id),
+            additional_params=additional_params,
+            limit=limit,
+            marker=marker,
+            fields=fields,
+            return_full_pages=False
```
Comment (anthonywee) [inline boxsdk/object/retention_policy.py]: trailing comma

File: boxsdk/pagination/page.py [hunk 24]
```diff
@@ -38,7 +39,7 @@ def _translator(self):
 
         :rtype:   :class:`Translator`
         """
-        return self._session.translator
+        return Translator()
```
Comment (anthonywee) [inline boxsdk/pagination/page.py]: Why can't we use the session's translator? Is it okay to create a new one every time here? Maybe a comment could help?

File: test/unit/client/test_client.py [hunk 25]
```diff
@@ -324,3 +325,47 @@ def test_create_enterprise_user_returns_the_correct_user_object(mock_client, moc
     assert isinstance(new_user, User)
     assert new_user.object_id == 1234
     assert new_user.name == test_user_name
+
```
Comment (anthonywee) [inline test/unit/client/test_client.py]: white space

File: test/unit/client/test_client.py [hunk 26]
```diff
@@ -324,3 +325,47 @@ def test_create_enterprise_user_returns_the_correct_user_object(mock_client, moc
...
+    policy_name = 'Test Retention Policy'
+    policy_type = 'indefinite'
+    disposition_action = 'remove_retention'
+    expected_url = mock_box_session.get_url('retention_policies')
+    expected_data = {
+        'policy_name': policy_name,
+        'policy_type': policy_type,
+        'disposition_action': disposition_action,
+        'can_owner_extend_retention': False,
+        'are_owners_notified': False
```
Comment (anthonywee) [inline test/unit/client/test_client.py]: trailing comma

File: test/unit/client/test_client.py [hunk 27]
```diff
@@ -324,3 +325,47 @@ def test_create_enterprise_user_returns_the_correct_user_object(mock_client, moc
...
+    assert policy.id == mock_policy['id']
+    assert policy.type == mock_policy['type']
+
+
+def test_get_retention_policies(mock_client, mock_box_session):
+    expected_url = mock_box_session.get_url('retention_policies')
+    mock_policy = {
+        'type': 'retention_policy',
+        'id': '12345',
+        'name': 'Test Retention Policy'
```
Comment (anthonywee) [inline test/unit/client/test_client.py]: trailing comma

File: test/unit/client/test_client.py [hunk 28]
```diff
@@ -324,3 +325,47 @@ def test_create_enterprise_user_returns_the_correct_user_object(mock_client, moc
...
+    }
+    mock_box_session.get.return_value.json.return_value = {
+        'limit': 100,
+        'entries': [mock_policy],
+        'next_marker': 'testMarker'
```
Comment (anthonywee) [inline test/unit/client/test_client.py]: trailing comma

File: test/unit/object/test_retention_policy.py [hunk 29]
```diff
@@ -0,0 +1,81 @@
+from __future__ import unicode_literals
```
Comment (anthonywee) [inline test/unit/object/test_retention_policy.py]: absolute_import

File: test/unit/object/test_retention_policy_assignment.py [hunk 30]
```diff
@@ -0,0 +1,15 @@
+from __future__ import unicode_literals
```
Comment (anthonywee) [inline test/unit/object/test_retention_policy_assignment.py]: absolute_import

File: boxsdk/object/retention_policy_assignment.py [hunk 31]
```diff
@@ -0,0 +1,11 @@
+# coding: utf-8
+
+from __future__ import unicode_literals
```
Comment (anthonywee) [inline boxsdk/object/retention_policy_assignment.py]: absolute_import

**Pull request merged**